### PR TITLE
os/bluestore/bluefs: Cleanup allocation consistency check code

### PR DESCRIPTION
**Pull request by aclamk**
When bluefs_log_replay_check_allocations is set _replay performs checks if
allocations/deallocations of extents are valid and properly aligned.
The changes include:
- now deallocations are also checked for valid alignment (more sanity checks)
- bluefs log read from super is checked once - before reading any transaction

Signed-off-by: Adam Kupczyk <akupczyk@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -996,6 +1001,39 @@ int BlueFS::_check_new_allocations(const bluefs_fnode_t& fnode,
   return 0;
 }
 
+int BlueFS::_check_delete_allocations(const bluefs_fnode_t& fnode,
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc:1004]: Am I right that _check_new_allocations and _check_delete_allocations differ in the following lines only?
if (!bs.test(pos)) {
  fail = true;
}
bs.reset(pos);
vs.
if (bs.test(pos)) {
  fail = true;
}
bs.set(pos);

If so may be simplify that and leave a single _check_allocations() function getting 'expected' parameter:
if (bs.test(pos) != expected) {
  fail = true;
}
bs.flip(pos);

Comment (aclamk) [reply src/os/bluestore/BlueFS.cc]: True. Those functions are complementary.
And you are right - melding them together improved readability.

**Comment by tchaikov** [comment]
this change conflicts with #42734 which is still referencing `first_log_check`

**Comment by ifed01** [comment]
> this change conflicts with #42734 which is still referencing `first_log_check`

#42734 is legacy now, #42750 supersedes it. The latter would rather conflict with the current PR as well so I think we should either handle them one-by-one  or ask @aclamk  to make a single PR from both..

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2021-08-13_12:30:46-rados-wip-kefu-testing-2021-08-13-1658-distro-basic-smithi/

cephadm failures are tracked by https://tracker.ceph.com/issues/52279

**Pull request merged**